### PR TITLE
[MachineVerifier] Query TargetInstrInfo for PHI nodes.

### LINKED ISSUE #108844: [SPIR-V] OpPhi not handled as a phi node by the MIR verifier
When expensive checks are enabled, MIR is checked after each pass.
SPIR-V has an OpPhi instruction, similar to LLVM's phi instruction.
But because the MIR verifier doesn't know about it, it expects all operands to be defined by a dominating block (unlike the phi instruction).

Might want to add a new bit in the td files (akin to isBarrier bit) so the verifier can know "this instruction is allowed to have partially undefined operands)?

### PR DESCRIPTION
**Pull request by Keenuts**
MachineVerifier can run post-ISel, and the SPIR-V can emit phi nodes. This means G_PHI/PHI are not the only 2 opcodes representing phi nodes. In addition, the SPIR-V instruction operands are ordered slightly differently (1 additional operand before the pairs).

There are multiple ways to solve this, but here I decided to go with the less invasive, but less generic method. However the refactoring mentioned in the rest of this commit can still be done later, as it also relies on TII exposing an `isPhiInstr`.

Alternative designs
===================

1st alternative:
----------------

Add a bit in MCInstDesc for phi instructions (See #110019). This was refused as the MCInstDesc bits are "pricy", and a that only impacts 3 instructions is not a wise expense.
Additionally, it doesn't help handling the changing operands layout, and still requires some additional info from the target to do so.

2nd alternative:
----------------

Refactorize `MachineInstr::isPHI()` to use `TargetInstrInfo`. This was almost possible, as `MachineInstr` often has a parent MBB, and thus a parent MF which links to the TargetInstrInfo.

In MachineInstr: `this->getMF().getTargetInstrInfo().isPhiInstr(*this)`

This however breaks as soon as the `MachineInstr` is removed from it's parent block.
Solving this requires passing TII as a parameter to isPHI. Passing TII requires each code using `isPHI` to also pass TII. This is a very invasive change, which impacts almost every backends, and several passes.

Fixes #108844

### CODE REVIEW

File: llvm/include/llvm/CodeGen/TargetInstrInfo.h [hunk 1]
```diff
@@ -2278,6 +2278,32 @@ class TargetInstrInfo : public MCInstrInfo {
...
+  }
+
+  // Returns the |index|'th [Value, Src] pair of this phi instruction.
+  virtual std::pair<MachineOperand, MachineBasicBlock *>
+  getPhiIncomingPair(const MachineInstr &MI, unsigned index) const {
+    // Behavior for G_PHI/PHI instructions.
+    MachineOperand Operand = MI.getOperand(index * 2 + 1);
+    MachineBasicBlock *MBB = MI.getOperand(index * 2 + 2).getMBB();
+    return {Operand, MBB};
+  }
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/TargetInstrInfo.h:2305]: I don't think the special case SPIRV operator should be treated like a generic phi like this (at least, not in any actual codegen uses). Can you attach sample failing MIR cases to the issue (or better yet, add machine verifier tests to this PR)? I think we'd be better off only touching something visible to the verifier, or otherwise avoiding clearing the NoPhis property for spirv

**Review by arsenm — COMMENTED** [review]
Ne

File: llvm/include/llvm/CodeGen/TargetInstrInfo.h (hunk 1 above)
Comment (arsenm) [reply llvm/include/llvm/CodeGen/TargetInstrInfo.h:2305]: I meant tests that hit the error in test/MachineVerifier, not tests that you are trying to make not fail the verifier

Comment (Keenuts) [reply llvm/include/llvm/CodeGen/TargetInstrInfo.h:2305]: > I don't think the special case SPIRV operator should be treated like a generic phi like this (at least, not in any actual codegen uses).

Why that? It's also a PHI node, exact same behavior & requirements (beginning of a BB, 1 pair by predecessor, only labels as predecessor argument, etc)

Comment (Keenuts) [reply llvm/include/llvm/CodeGen/TargetInstrInfo.h:2305]: > I meant tests that hit the error in test/MachineVerifier, not tests that you are trying to make not fail the verifier

Added a new MachineVerifer test: `llvm/test/MachineVerifier/verifier-phi-spirv.mir`

File: llvm/test/MachineVerifier/verifier-phi-spirv.mir [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# RUN: llc -o - %s -mtriple=spirv-unknown-vulkan-compute -verify-machineinstrs -run-pass=none | FileCheck %s
+# REQUIRES: spirv-registered-target
```
Comment (arsenm) [inline llvm/test/MachineVerifier/verifier-phi-spirv.mir:2]: Better to move to a new test/MachineVerifier/SPIRV

File: llvm/test/MachineVerifier/verifier-phi-spirv.mir [hunk 3]
```diff
@@ -0,0 +1,31 @@
+# RUN: llc -o - %s -mtriple=spirv-unknown-vulkan-compute -verify-machineinstrs -run-pass=none | FileCheck %s
```
Comment (arsenm) [inline llvm/test/MachineVerifier/verifier-phi-spirv.mir:1]: Don't need -verify-machineinstrs, the parser always verifies

File: llvm/include/llvm/CodeGen/TargetInstrInfo.h (hunk 1 above)
Comment (arsenm) [reply llvm/include/llvm/CodeGen/TargetInstrInfo.h:2305]: > exact same behavior & requirements

It's not the same. PHI requires an entry for each incoming edge, and OpPhi requires one entry per predecessor (this is another case where it will likely be easier to conform the SPIRV operation to the generic's form, and then just fix up the encoding f

**Review by arsenm — COMMENTED** [review]
I'm still inclined to think it will be easier to just select to PHI and only treat the conversion to OpPhi as a late final encoding detail. 

The ordinary compilation flow emits the PHI, which is later destructured to copies at the start of register allocation. The SPIRV analog would be to adjust at a similarly late point to something closer to OpPhi 

The point of the backend is to emit SPIRV at the end, that doesn't require you to turn MIR into SPIRV. It's a different representation

**Comment by VyacheslavLevytskyy** [comment]
We should not say that `OpPhi` is not a `phi` just because it has one difference that is subtle enough to make us spend a lot of time trying to create a reproducer for it. This difference is buried deep and doesn’t interfere with anything. Just the opposite this is an argument against the suggested approach of delayed translation of phi to OpPhi, as it would make a price of delayed translation higher, along with SPIRV types and, in general, violation of the instruction selection notion.

We may describe the issue so that GenCode doesn’t expect that after lowering a target still supports a higher-level opcode representing a PHI node. SPIR-V has a PHI opcode, and it doesn’t match with what GenCode expects from a lowered code. I think that this PR presents a normal engineering practice of using defaults and overrides to resolve this issue.

**Comment by VyacheslavLevytskyy** [comment]
In fact https://github.com/llvm/llvm-project/issues/108844 might have been worded as follows:
1. By definition [https://llvm.org/docs/CodeGenerator.html#instruction-selection-section] Instruction Selection results in target-specific machine instructions. It's the SPIR-V code set in our case.
2. According to the SPIR-V spec, OpPhi [https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpPhi] is used to implement the φ node in the SSA graph representing the function.
3. When Machine Verifier reports `Virtual register defs don't dominate all uses` applied to OpPhi, it is an issue in implementation of Machine Verifier, because this is reported **after** Instruction Selection and the SPIR-V spec considers this use valid.

In other words, it doesn't matter what is the difference between phi and OpPhi when we discuss this issue. Machine Verifier may not consider this a bad code, because Instruction Selection results in *target-specific* instructions, and by the SPIR-V spec (target-specific) OpPhi implements the φ node.

**Review by michalpaszkowski — APPROVED** [review]
After the discussions at the LLVM Developers' Meeting and general vision other contributors have about GlobalISel, I think this still might be the best way forward. Workarounds in the code gen pipeline significantly contribute both to logical complexity and compilation speed. Also, from a purist point of view, selected instructions should not require any further "selection". Using TII is a reasonable flexibility that in this and other cases may help other targets in the future. Target specific information is used in other cases in MachineVerifier.

I am open to discuss this (e.g. at our next SPIR-V backend meeting on Wednesday next week), but overall the change is justified and rather helps.

File: llvm/include/llvm/CodeGen/TargetInstrInfo.h [hunk 4]
```diff
@@ -2278,6 +2278,32 @@ class TargetInstrInfo : public MCInstrInfo {
     llvm_unreachable("unknown number of operands necessary");
   }
 
+  // This this instruction a PHI node.
```
Comment (VyacheslavLevytskyy) [inline llvm/include/llvm/CodeGen/TargetInstrInfo.h:2281]: `This this`

File: llvm/include/llvm/CodeGen/TargetInstrInfo.h [hunk 5]
```diff
@@ -2278,6 +2278,32 @@ class TargetInstrInfo : public MCInstrInfo {
...
+           MI.getOpcode() == TargetOpcode::PHI;
+  }
+
+  // Returns the number of [Value, Src] pairs this phi instruction has.
+  // Only valid to call on a phi node.
+  virtual unsigned getNumPhiIncomingPair(const MachineInstr &MI) const {
+    assert(isPhiInstr(MI));
+    // G_PHI/PHI only have a single operand before the pairs. 2 Operands per
+    // pair.
+    return (MI.getNumOperands() - 1) / 2;
```
Comment (VyacheslavLevytskyy) [inline llvm/include/llvm/CodeGen/TargetInstrInfo.h:2295]: or `>> 1` ?

File: llvm/lib/CodeGen/MachineVerifier.cpp [hunk 6]
```diff
@@ -3211,17 +3211,22 @@ void MachineVerifier::calcRegsRequired() {
         todo.insert(Pred);
     }
 
-    // Handle the PHI node.
-    for (const MachineInstr &MI : MBB.phis()) {
-      for (unsigned i = 1, e = MI.getNumOperands(); i != e; i += 2) {
+    // Handle the PHI nodes.
+    // Note: MBB.phis() only returns range containing PHI/G_PHI instructions.
+    // MachineVerifier checks MIR post-ISel, and some backends do have their own
```
Comment (VyacheslavLevytskyy) [inline llvm/lib/CodeGen/MachineVerifier.cpp:3216]: Maybe instead "and some backends do have their own phi nodes" something like "and a backend may have a phi node instruction in the instruction set (e.g., SPIR-V backend)". The point of the suggestion is that I'm not aware of another backend with a phi node instruction.

File: llvm/lib/Target/SPIRV/SPIRVInstrInfo.cpp [hunk 7]
```diff
@@ -267,3 +267,27 @@ bool SPIRVInstrInfo::expandPostRAPseudo(MachineInstr &MI) const {
...
+
+// The SPIR-V backends can emit the OpPhi instruction.
+bool SPIRVInstrInfo::isPhiInstr(const MachineInstr &MI) const {
+  return TargetInstrInfo::isPhiInstr(MI) || MI.getOpcode() == SPIRV::OpPhi;
+}
+
+unsigned SPIRVInstrInfo::getNumPhiIncomingPair(const MachineInstr &MI) const {
+  // OpPhi has 2 operands before the [Value, Src] pairs.
+  if (MI.getOpcode() == SPIRV::OpPhi)
+    return (MI.getNumOperands() - 2) / 2;
```
Comment (VyacheslavLevytskyy) [inline llvm/lib/Target/SPIRV/SPIRVInstrInfo.cpp:279]: `>> 1`?

**Comment by Keenuts** [comment]
Thanks for the reviews!
@arsenm @michalpaszkowski : I wasn't at the dev meeting, addressed the review comments (and also added the test for the duplicated predecessor), but unsure what's the status on the approval.

**Review by VyacheslavLevytskyy — APPROVED** [review]
